### ISSUE TITLE
A refactoring idea for errors

### ISSUE THREAD

**Issue by eriksvedang** [carp-lang/Carp#1234]
How about adding a typeclass that we implement for all the different error types in the compiler. It would have two methods, one for producing the long nice message (currently done using `show`) and one that emits the machine readable version of the same error.

This would ensure that we can always get both versions, which currently is not true. This leads to inconsistencies in error reporting, where some kinds of errors can't be machine readable. 

It could also make some code less tied to specific kinds of errors, and thus more generic and reusable.

**Comment by scolsen** [comment]
Sounds good!

I think one thing we'll want to consider is whether or not, especially for machine readable strings, the class should interface against XObjs or Info, e.g. do we want:

```haskell
class Error a where
  humanReadable :: a -> String
  machineReadable :: a -> String
```

Or

```haskell
class Error a where
  humanReadable :: a -> String
  machineReadable :: a -> Info -> String
```

Or something similar.

**Comment by eriksvedang** [comment]
Yeah, it should take `Info` since it's necessary to emit the right info. Same goes for `ExecutionMode` and `filePathPrintLength` I believe. So maybe it should just take `Context` for that?
